hail/src/main/scala/is/hail/io/fs/TerraAzureStorageFS.scala

```diff
@@ -0,0 +1,93 @@
+package is.hail.io.fs
+
+import scala.collection.mutable
+
+import is.hail.shadedazure.com.azure.core.credential.{AzureSasCredential, TokenCredential, TokenRequestContext}
+import is.hail.shadedazure.com.azure.identity.{DefaultAzureCredential, DefaultAzureCredentialBuilder}
+import is.hail.utils._
+
+import org.apache.http.client.utils.URIBuilder
+import org.apache.http.HttpEntity
+import org.apache.http.client.methods.HttpPost
+import org.apache.http.entity.{ContentType, StringEntity}
+import org.apache.http.impl.client.HttpClients
+import org.apache.http.params.HttpParams
+import org.apache.http.util.EntityUtils
+import org.apache.log4j.Logger
+import org.json4s.{DefaultFormats, Formats}
+import org.json4s.jackson.JsonMethods
+
+
+object TerraAzureStorageFS {
+  private val workspaceManagerUrl = sys.env.get("WORKSPACE_MANAGER_URL")
+  private val workspaceId = sys.env.get("WORKSPACE_ID")
+  private val containerResourceId = sys.env.get("WORKSPACE_STORAGE_CONTAINER_ID")
+  private val storageContainerUrl = sys.env.get("WORKSPACE_STORAGE_CONTAINER_URL")
+  private val log = Logger.getLogger(getClass.getName)
+}
+
+class TerraAzureStorageFS() extends FS {
+  type URL = AzureStorageFSURL
+  import TerraAzureStorageFS.{log, workspaceManagerUrl, workspaceId, containerResourceId, storageContainerUrl}
+
+  private[this] val azureFS = new AzureStorageFS()
+  private[this] val credential: DefaultAzureCredential = new DefaultAzureCredentialBuilder().build()
+  private[this] val httpClient = HttpClients.custom().build()
+  private[this] val sasTokenCache = mutable.Map[String, (String, Long)]()
+
+  def parseUrl(filename: String): AzureStorageFSURL = {
+    val urlStr = storageContainerUrl match {
```

Comment:
What does it mean to not have a storage container in Azure?

Reply:
We should always have one if we're getting this far. I'll turn this into a non-optional field of the class.